HRM Virtual Host Machine Deployment Failure for  a SingleRegion-DualAZ VCF Instance

### Code of Conduct

- [X] I have read and agree to the [Code of Conduct](https://github.com/vmware/power-validated-solutions-for-cloud-foundation/blob/main/CODE_OF_CONDUCT.md).
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### VMware Cloud Foundation

5

### Module Version

2.11.0.1038

### PowerShell Version

 7.4.3

### PowerCLI Version

13.2.1.22851661

### PowerVCF Version

2.4.1.1000

### Guest Operating System

Windows 2019

### Environment Details

_No response_

### Description

I am deploying HRM on a single-region and dual AZ VCF instance. The end-to-end deployment failed. It seems there is a bug in the function Deploy-PhotonAppliance. After investigation, I found the function can't get the right portgroup for a stretched cluster: it got the portgroup from the witness node. 

  


### Error or Debug Output

https://gist.github.com/insidepacket/12a231b5b502bfaa2e34930576445062

### Expected Behavior

PT would deploy Virtual Host Appliance successfully.

### Actual Behavior

The virtual host appliance deployment failed with the following error: [ERROR] Deploying and Powering On Virtual Machine (sfo-m01-hrm01) in vCenter Server (sfo-m01-vc01.sfo.rainpole.io): POST_VALIDATION_FAILED


### Steps to Reproduce

Step 1: Deployed a Single region dual AZ VCF instance by use of PT
Step 2: Install WS1 and vROPS in vRSLCM
Step 3: Start PT, then select option 50. VMware Validated Solutions (Sub-Menu)
Step 4: Select option 10 (HRM) Health Reporting and Monitoring
Step 5: Run options 01 and 02 to prepare the implementation
Step 6:  Run option 5 (End-to-End) Deployment. You will be able to hit the issue.

### Log Fragments and Files

_No response_

### Screenshots

_No response_

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@
 - Enhanced `Get-LocalAccountLockout` cmdlet to handle "null" values got as "Not Configured" when appliance is not configured.
 - Enhanced `Export-DriJsonSpec` cmdlet to define the path to the Kubectl utility.
 - Enhanced `Start-DriMenu` cmdlet to use the new JSON value for the path to the Kubectl utility.
+- Enhanced `Deploy-PhotonAppliance` cmdlet with an input parameter for the portgroup to attach the Virtual Machine too.
 
 ## v2.11.0
 

--- a/PowerValidatedSolutions.psd1
+++ b/PowerValidatedSolutions.psd1
@@ -11,7 +11,7 @@
     RootModule = 'PowerValidatedSolutions.psm1'
 
     # Version number of this module.
-    ModuleVersion = '2.11.1.1003'
+    ModuleVersion = '2.11.1.1004'
 
     # ID used to uniquely identify this module
     GUID              = 'a6dfed7b-65d2-4da2-bdcc-7f3d3df9b75d'

--- a/PowerValidatedSolutions.psm1
+++ b/PowerValidatedSolutions.psm1
@@ -26556,6 +26556,7 @@ Function Export-HrmJsonSpec {
                 'sddcManagerUser'              = $pnpWorkbook.Workbook.Names["sso_default_admin"].Value
                 'sddcManagerPass'              = $pnpWorkbook.Workbook.Names["administrator_vsphere_local_password"].Value
                 'mgmtSddcDomainName'           = $pnpWorkbook.Workbook.Names["mgmt_sddc_domain"].Value
+                'mgmtVmPortgroup'              = $pnpWorkbook.Workbook.Names["mgmt_cl01_az1_mgmt_vm_pg"].Value
                 'vmFolder'                     = $pnpWorkbook.Workbook.Names["hrm_vm_folder"].Value
                 'vmName'                       = $pnpWorkbook.Workbook.Names["hrm_vm_hostname"].Value
                 'fqdn'                         = $pnpWorkbook.Workbook.Names["hrm_vm_fqdn"].Value
@@ -26579,6 +26580,7 @@ Function Export-HrmJsonSpec {
             if ($pnpWorkbook.Workbook.Names["mgmt_stretched_cluster_chosen"].Value -eq "Include") {
                 $jsonObject | Add-Member -notepropertyname 'drsVmGroupNameAz' -notepropertyvalue $pnpWorkbook.Workbook.Names["mgmt_az1_vm_group_name"].Value
             }
+
             Close-ExcelPackage $pnpWorkbook -NoSave -ErrorAction SilentlyContinue
             $jsonObject | ConvertTo-Json -Depth 12 | Out-File -Encoding UTF8 -FilePath $jsonFile
             $jsonInput = (Get-Content -Path $jsonFile) | ConvertFrom-Json
@@ -26703,7 +26705,7 @@ Function Invoke-HrmDeployment {
 
                     if (!$failureDetected) {
                         Show-PowerValidatedSolutionsOutput -message "Deploying the Host Virtual Machine for $solutionName"
-                        $StatusMsg = Deploy-PhotonAppliance -server $jsonInput.sddcManagerFqdn -user $jsonInput.sddcManagerUser -pass $jsonInput.sddcManagerPass -sddcDomain $jsonInput.mgmtSddcDomainName -domain $jsonInput.searchDomain -hostname $jsonInput.vmName -ipAddress $jsonInput.ipAddress -netmask $jsonInput.netmask -gateway $jsonInput.gateway -dnsServer $jsonInput.dns -ntpServer $jsonInput.ntp -adminPassword $jsonInput.adminPassword -enableSsh True -folder $jsonInput.vmFolder -ovaPath ($binaries + $jsonInput.ova) -WarningAction SilentlyContinue -ErrorAction SilentlyContinue -WarningVariable WarnMsg -ErrorVariable ErrorMsg
+                        $StatusMsg = Deploy-PhotonAppliance -server $jsonInput.sddcManagerFqdn -user $jsonInput.sddcManagerUser -pass $jsonInput.sddcManagerPass -sddcDomain $jsonInput.mgmtSddcDomainName -domain $jsonInput.searchDomain -hostname $jsonInput.vmName -portgroup $jsonInput.mgmtVmPortgroup -ipAddress $jsonInput.ipAddress -netmask $jsonInput.netmask -gateway $jsonInput.gateway -dnsServer $jsonInput.dns -ntpServer $jsonInput.ntp -adminPassword $jsonInput.adminPassword -enableSsh True -folder $jsonInput.vmFolder -ovaPath ($binaries + $jsonInput.ova) -WarningAction SilentlyContinue -ErrorAction SilentlyContinue -WarningVariable WarnMsg -ErrorVariable ErrorMsg
                         messageHandler -statusMessage $StatusMsg -warningMessage $WarnMsg -errorMessage $ErrorMsg; if ($ErrorMsg) { $failureDetected = $true }
                     }
 
@@ -26845,8 +26847,8 @@ Function Deploy-PhotonAppliance {
         - Deploys the Photon Appliance into a vSphere Cluster
 
         .EXAMPLE
-        Deploy-PhotonAppliance -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -sddcDomain sfo-m01 -hostname sfo-m01-hrm01 -ipAddress 172.18.95.50 -netmask "24 (255.255.255.0)" -gateway 172.18.95.1 -domain sfo.rainpole.io -dnsServer "172.18.95.4 172.18.95.5" -ntpServer ntp.sfo.rainpole.io -adminPassword VMw@re1! -enableSsh True -folder sfo-m01-fd-hrm -ovaPath .\vvs-appliance-v0.1.0.ova
-        This example deploys the Photon appliance named sfo-m01-hrm01.
+        Deploy-PhotonAppliance -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -sddcDomain sfo-m01 -hostname sfo-m01-hrm01 -portgroup sfo-m01-cl01-vds01-pg-vm-mgmt -ipAddress 172.18.95.50 -netmask "24 (255.255.255.0)" -gateway 172.18.95.1 -domain sfo.rainpole.io -dnsServer "172.18.95.4 172.18.95.5" -ntpServer ntp.sfo.rainpole.io -adminPassword VMw@re1! -enableSsh True -folder sfo-m01-fd-hrm -ovaPath .\vvs-appliance-v0.1.0.ova
+        This example deploys the Photon appliance named sfo-m01-hrm01
 
         .PARAMETER server
         The fully qualified domain name of the SDDC Manager.
@@ -26862,6 +26864,9 @@ Function Deploy-PhotonAppliance {
 
         .PARAMETER hostname
         The hostname of the Photon appliance.
+
+        .PARAMETER portgroup
+        The portgroup to attach to the Photon appliance.
 
         .PARAMETER domain
         The domain namespace for the Photon appliance.
@@ -26900,6 +26905,7 @@ Function Deploy-PhotonAppliance {
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$sddcDomain,
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$hostname,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$portgroup,
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$domain,
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$ipAddress,
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$netmask,
@@ -26923,7 +26929,6 @@ Function Deploy-PhotonAppliance {
                         if (Test-VsphereConnection -server $($vcfVcenterDetails.fqdn)) {
                             if (Test-VsphereAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
                                 if (!(Get-VM -Name $hostname -Server ($vcfVcenterDetails.fqdn).Name -ErrorAction Ignore)) {
-                                    $portgroup = ((Get-VMHost)[0] | Get-VMHostNetwork | Select-Object Hostname, VMkernelGateway -ExpandProperty VirtualNic | where-object { $_.DeviceName -eq "vmk0" }).PortGroupName
                                     $cluster = (Get-VCFCluster | Where-Object { $_.id -eq ((Get-VCFWorkloadDomain | Where-Object { $_.name -eq $sddcDomain }).clusters.id) }).Name
                                     $datastore = (Get-VCFCluster | Where-Object { $_.id -eq ((Get-VCFWorkloadDomain | Where-Object { $_.name -eq $sddcDomain }).clusters.id) }).primaryDatastoreName
                                     $datacenter = (Get-Datacenter -Cluster $cluster).Name

--- a/docs/documentation/functions/solutions/hrm/Deploy-PhotonAppliance.md
+++ b/docs/documentation/functions/solutions/hrm/Deploy-PhotonAppliance.md
@@ -7,7 +7,7 @@ Deploy a Photon appliance.
 ## Syntax
 
 ```powershell
-Deploy-PhotonAppliance [-server] <String> [-user] <String> [-pass] <String> [-sddcDomain] <String> [-hostname] <String> [-domain] <String> [-ipAddress] <String> [-netmask] <String> [-gateway] <String> [-dnsServer] <String> [-ntpServer] <String> [-adminPassword] <String> [-enableSsh] <String> [-folder] <String> [[-ovaPath] <String>] [-ProgressAction <ActionPreference>] [<CommonParameters>]
+Deploy-PhotonAppliance [-server] <String> [-user] <String> [-pass] <String> [-sddcDomain] <String> [-hostname] <String> [-portgroup] <String> [-domain] <String> [-ipAddress] <String> [-netmask] <String> [-gateway] <String> [-dnsServer] <String> [-ntpServer] <String> [-adminPassword] <String> [-enableSsh] <String> [-folder] <String> [[-ovaPath] <String>] [<CommonParameters>]
 ```
 
 ## Description
@@ -24,7 +24,8 @@ The cmdlet connects to SDDC Manager using the -server, -user, and -password valu
 ### Example 1
 
 ```powershell
-Deploy-PhotonAppliance -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -sddcDomain sfo-m01 -hostname sfo-m01-hrm01 -ipAddress 172.18.95.50 -netmask "24 (255.255.255.0)" -gateway 172.18.95.1 -domain sfo.rainpole.io -dnsServer "172.18.95.4 172.18.95.5" -ntpServer ntp.sfo.rainpole.io -rootPassword VMw@re1! -enableSsh True -folder sfo-m01-fd-hrm -ovaPath .\vvs-appliance-v0.1.0.ova
+Deploy-PhotonAppliance -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -sddcDomain sfo-m01 -hostname sfo-m01-hrm01 -portgroup sfo-m01-cl01-vds01-pg-vm-mgmt -ipAddress 172.18.95.50 -netmask "24 (255.255.255.0)" -gateway 172.18.95.1 -domain sfo.rainpole.io -dnsServer "172.18.95.4 172.18.95.5" -ntpServer ntp.sfo.rainpole.io -adminPassword VMw@re1! -enableSsh True -folder sfo-m01-fd-hrm -ovaPath .\vvs-appliance-v0.1.0.ova
+This example deploys the Photon appliance named sfo-m01-hrm01
 ```
 
 This example deploys the Photon appliance named sfo-m01-hrm01.
@@ -106,6 +107,22 @@ Aliases:
 
 Required: True
 Position: 5
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -portgroup
+
+The portgroup to attach to the Photon appliance.
+
+```yaml
+Type: String
+Parameter Sets: (All)
+Aliases:
+
+Required: True
+Position: 6
 Default value: None
 Accept pipeline input: False
 Accept wildcard characters: False

